File: src/test/scala/com/johnsnowlabs/nlp/annotators/classifier/dl/AlbertForMultipleChoiceTest.scala
package com.johnsnowlabs.nlp.annotators.classifier.dl

import com.johnsnowlabs.nlp.{Annotation, AssertAnnotations, MultiDocumentAssembler}
import com.johnsnowlabs.nlp.annotators.SparkSessionTest
import com.johnsnowlabs.nlp.base.LightPipeline
import com.johnsnowlabs.tags.SlowTest
import org.apache.spark.ml.Pipeline
import org.scalatest.flatspec.AnyFlatSpec

class AlbertForMultipleChoiceTest extends AnyFlatSpec with SparkSessionTest {

  import spark.implicits._

  lazy val pipelineModel = getAlbertForMultipleChoicePipelineModel

  val testDataframe =
    Seq(("The Eiffel Tower is located in which country?", "Germany, France, Italy"))
      .toDF("question", "context")

  "AlbertForMultipleChoice" should "answer a multiple choice question" taggedAs SlowTest in {
    val resultDf = pipelineModel.transform(testDataframe)
    resultDf.show(truncate = false)

    val result = AssertAnnotations.getActualResult(resultDf, "answer")
    result.foreach { annotation =>
      annotation.foreach(a => assert(a.result.nonEmpty))
    }
  }

  it should "work with light pipeline fullAnnotate" taggedAs SlowTest in {
    val lightPipeline = new LightPipeline(pipelineModel)
    val resultFullAnnotate = lightPipeline.fullAnnotate(
      "The Eiffel Tower is located in which country?",
      "Germany, France, Italy")
    println(s"resultAnnotate: $resultFullAnnotate")

    val answerAnnotation = resultFullAnnotate("answer").head.asInstanceOf[Annotation]

    assert(answerAnnotation.result.nonEmpty)
  }

  it should "work with light pipeline annotate" taggedAs SlowTest in {
    val lightPipeline = new LightPipeline(pipelineModel)
    val resultAnnotate = lightPipeline.annotate(
      "The Eiffel Tower is located in which country?",
      "Germany, France, Italy")
    println(s"resultAnnotate: $resultAnnotate")

    assert(resultAnnotate("answer").head.nonEmpty)
  }

  private def getAlbertForMultipleChoicePipelineModel = {
    val documentAssembler = new MultiDocumentAssembler()
      .setInputCols("question", "context")
      .setOutputCols("document_question", "document_context")

    val bertForMultipleChoice = AlbertForMultipleChoice
      .pretrained()
      .setInputCols("document_question", "document_context")
      .setOutputCol("answer")

    val pipeline = new Pipeline().setStages(Array(documentAssembler, bertForMultipleChoice))

    pipeline.fit(emptyDataSet)
  }

}

File: python/test/annotator/classifier_dl/albert_for_multiple_choice_test.py
#  Copyright 2017-2024 John Snow Labs
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import unittest

import pytest

from sparknlp.annotator.classifier_dl.albert_for_multiple_choice import AlbertForMultipleChoice
from sparknlp.base import *
from test.util import SparkContextForTest


class AlbertForMultipleChoiceTestSetup(unittest.TestCase):
    def setUp(self):

        sparkNLPModelPath = "/media/danilo/Data/Danilo/JSL/models/transformers/spark-nlp"

        self.spark = SparkContextForTest.spark
        self.question = "The Eiffel Tower is located in which country?"
        self.choices = "Germany, France, Italy"

        self.spark = SparkContextForTest.spark
        empty_df = self.spark.createDataFrame([[""]]).toDF("text")

        document_assembler = MultiDocumentAssembler() \
            .setInputCols(["question", "context"]) \
            .setOutputCols(["document_question", "document_context"])

        albert_for_multiple_choice = AlbertForMultipleChoice.load(sparkNLPModelPath + "/openvino/albert_multiple_choice_openvino") \
            .setInputCols(["document_question", "document_context"]) \
            .setOutputCol("answer")

        pipeline = Pipeline(stages=[document_assembler, albert_for_multiple_choice])

        self.pipeline_model = pipeline.fit(empty_df)


@pytest.mark.slow
class AlbertForMultipleChoiceTest(AlbertForMultipleChoiceTestSetup, unittest.TestCase):

    def setUp(self):
        super().setUp()
        self.data = self.spark.createDataFrame([[self.question, self.choices]]).toDF("question","context")
        self.data.show(truncate=False)

    def test_run(self):
        result_df = self.pipeline_model.transform(self.data)
        result_df.show(truncate=False)
        for row in result_df.collect():
            self.assertTrue(row["answer"][0].result != "")


@pytest.mark.slow
class LightAlbertForMultipleChoiceTest(AlbertForMultipleChoiceTestSetup, unittest.TestCase):

    def setUp(self):
        super().setUp()

    def runTest(self):
        light_pipeline = LightPipeline(self.pipeline_model)
        annotations_result = light_pipeline.fullAnnotate(self.question,self.choices)
        print(annotations_result)
        for result in annotations_result:
            self.assertTrue(result["answer"][0].result != "")

        result = light_pipeline.annotate(self.question,self.choices)
        print(result)
        self.assertTrue(result["answer"] != "")
